improver/psychrometric_calculations/significant_phase_mask.py

```diff
@@ -0,0 +1,135 @@
+#
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Module for calculating the significant phase mask."""
+
+import numpy as np
+
+from improver import BasePlugin
+from improver.metadata.utilities import (
+    create_new_diagnostic_cube,
+    generate_mandatory_attributes,
+)
+
+
+class SignificantPhaseMask(BasePlugin):
+    """
+    Derives a mask (1 and 0) for the specified precipitation phase indicating whether
+    that phase is the dominant phase at each point based on input snow-fraction data.
+    The decision is: snow-fraction <= 0.01: Rain; snow-fraction >= 0.99: Snow; Sleet
+    in between.
+    """
+
+    def __init__(self):
+        """Initialise the class"""
+        self.lower_threshold = 0.01
+        self.upper_threshold = 0.99
+        self.phase_operator = {
+            "rain": self._rain_phase,
+            "sleet": self._sleet_phase,
+            "snow": self._snow_phase,
+        }
+
+    @staticmethod
+    def _validate_snow_fraction(snow_fraction):
+        """Ensures that the input snow-fraction field has appropriate name
+        (snow_fraction), units (1) and data (between 0 and 1 inclusive).
+
+        Raises
+            ValueError:
+                If any of the above are not True.
+        """
+        if snow_fraction.name() != "snow_fraction":
+            raise ValueError(
+                f"Expected cube named 'snow_fraction', not {snow_fraction.name()}"
+            )
+        if f"{snow_fraction.units}" != "1":
+            raise ValueError(f"Expected cube with units '1', not {snow_fraction.units}")
+        if (
+            isinstance(snow_fraction.data, np.ma.masked_array)
+            and snow_fraction.data.mask.any()
+        ):
+            raise NotImplementedError("SignificantPhaseMask cannot handle masked data")
```

Comment:
Yes. Kind of.
The reason I needed to make the snow-fraction plugin handle masked data is because some of the rain and snow data we use contain masked arrays, but with no masked points. This changed the behaviour of the infilling and gave an unexpected result. It seemed obvious to me what the behaviour of the snow-fraction plugin should be for masked data (where masked-True in either input, output should also be masked-True - all other points should have a valid number either from main calculation or from filling in divide-by-zero points). It is not obvious what the behaviour in the phase-mask plugin should be because it states that we must have a significant phase at every point which would not be true if any point was masked-True. I could change the snow-fraction code to reject any fields with masked-True values, but it didn't seem necessary.